plugins/database/mysql/connection_producer.go

```diff
@@ -142,12 +164,43 @@ func (c *mySQLConnectionProducer) Connection(ctx context.Context) (interface{},
 		c.db.Close()
 	}
 
+	driverName := driverMySQL
+	if c.isCloud {
+		driverName = c.cloudDriverName
+
+		//@TODO - move these to init?
+		credentials := c.RawConfig["credentials"]
+
+		// for _most_ sql databases, the driver itself contains no state. In the case of google's cloudsql drivers,
+		// however, the driver might store a credentials file, in which case the state stored by the driver is in
+		// fact critical to the proper function of the connection.
+		//
+		// This poses one big obvious problem - each configured cloud database might/will need its own driver registration,
+		// the name of which we have to track, and even worse in the MySQL case, that name is also the name of the
+		// dialer that is registered by MySQL in the dsn: protocol in the DSN acts double duty if a custom dialer
+		// is registered. This means that either we can't hide the name of the dialer from the user OR we have to rewrite
+		// the DSN after the user provides it. We already do this, KIND OF for the TLS config, but this modification
+		// is much more dramatic.
+		_, err := registerDriverMySQL(driverName, credentials.(string))
+		if err != nil {
+			return nil, err
+		}
+
+		//@TODO store driver cleanup
+		//c.cleanupDrivers = append(c.cleanupDrivers, cleanupFunc)
+	}
+
 	connURL, err := c.addTLStoDSN()
 	if err != nil {
 		return nil, err
 	}
 
-	c.db, err = sql.Open("mysql", connURL)
+	cloudURL, err := c.rewriteProtocolForGCP(connURL)
```

Comment:
Will this return an error for non-CloudSQL DBs?

Reply:
in the version i juuuust wrote, i moved the isCloud check to the top of the function, so it should just noop if we're not in cloud mode.